.github/workflows/main.yaml

```diff
@@ -1,36 +1,205 @@
 #
 # Generated the prime numbers and generated used by SSH DH Group KEX.
 #
-name: GEN
+# Generating is fast. Max 15 minutes.
+# Validating is slow. 30 hours.
+#
+# Since the GitHub VM have 2 CPU each job will run 2 validating tasks.
+# A job can run for 6 hours... so 12 hours of validation time per job.
+# With 3 jobs for a certain size we should do it.
+#
+# It generates primes for 1024, 1536, 2048, 3072, 4096, 6144, 7680, 8192
+#
+name: CD
 
 # Only run on pull requests and only for Python files.
 on:
-  push:
-    branches: [ master ]
+  # Pull requests are enabled for development.
   pull_request:
-    branches: [ master ]
+    branches: [ main ]
 
+  # https://crontab.guru
+  schedule:
+    # At 01:00 on every 2 weeks on Saturday
+    - cron: '0 1 1-7,15-21 * 6'
 
 jobs:
 
+  prevent_duplicates:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: chevah/auto-cancel-redundant-job@v1
+    - uses: chevah/python-info-action@v1
+
+
   generate:
-    name: 'Size ${{ matrix.size }}'
+    name: 'ssh-keygen -M generate ${{ matrix.size }}'
+    needs: prevent_duplicates
     strategy:
-      fail-fast: false
       matrix:
-        # size: [ 1024, 1536, 2048, 3072, 4096, 6144, 8192]
-        size: [ 1024, 1536, 2048]
+        size: [ 1024, 1536,  2048, 3072, 4096 ]
+        split: [2]
+        include:
+          - size: 6144
+            split: 4
+          - size: 7680
+            split: 10
+          - size: 8192
+            split: 10
+
 
     runs-on: ubuntu-latest
     steps:
 
-    # Make sure we don't have multiple jobs
-    - uses: chevah/auto-cancel-redundant-job@v1
-
     - name: Candidate
       run: |
-        ssh-keygen -G candidate-${{ matrix.size }} -b ${{ matrix.size }}
+
+        ssh-keygen -M generate -O bits=${{ matrix.size }} full.1 &
+        PS1=$!
+        ssh-keygen -M generate -O bits=${{ matrix.size }} full.2 &
+        PS2=$!
+
+        wait $PS1 $PS2
+
+        cat full.2 >> full.1
+
+        # Split as round-robin as it looks like the first batch alwasy
+        # takes longer to validate.
+        split -d -a 1 -n r/${{ matrix.split }} full.1 g.
```

Comment:
Yes.... the workflow is a bit complicated :(

They are split so that screening for a single file/part will not take more than 6 hours.

I am trying to reduce the usage of our own buildbot and VM and outsource to GitHub.

At the same time, my goal here is to share some resources, so other people can pick weekly randomly generated moduli files, rather than rely on the upstream OpenSSH generated 3 times per year.

And we can't add (with decent security in place) our self-hosted VM to a public repo :(

Reply:
Not sure there would be interest in using moduli files from untrusted sources… People knowledgeable enough to audit this will probably not trust GitHub infrastructure. And people not knowledgeable enough don't care about the moduli files anyway. :-]

Maybe using a self-hosted GitHub runner that we have to run anyway, like `rhel-8`, would make it possible to have this simplified? It could run over a weekend, e.g. start Saturday evening every first week of the month or something. Or maybe once a year, on Jan 1st, if you ask me…

Reply:
Our last run which only covered half of the work took - 41 hrs ... so running the same job would take 80 hours (at mininum)

This is over the limit of 72 hours per job allowed by GitHub Actions.

So even when moved to our own VMs, if we switch to GitHub Actions we need to split the job to more than one VM.

-----------

In terms of security, right now I trust the GitHub self destroying VM more than our long-running VMs.

Our VMs can be compromised in the many months they are alive and we might not notice.

I think that many people trust GitHub Action  / Azure Pipeplines for they production deployment and usage.